Google.Solutions.Compute/Iap/SshRelayListener.cs

```diff
@@ -0,0 +1,184 @@
+        {
+            this.server = server;
+            this.LocalPort = localPort;
+
+            this.listener = new TcpListener(new IPEndPoint(IPAddress.Loopback, localPort));
+        }
+
+        //---------------------------------------------------------------------
+        // Events
+        //---------------------------------------------------------------------
+
+        private void OnClientConnected(string client)
+        {
+            if (this.ClientConnected != null)
+            {
+                this.ClientConnected(this, new ClientEventArgs(client));
+            }
+        }
+        private void OnClientDisconnected(string client)
+        {
+            if (this.ClientDisconnected != null)
+            {
+                this.ClientDisconnected(this, new ClientEventArgs(client));
+            }
+        }
+        private void OnConnectionFailed(Exception e)
+        {
+            if (this.ConnectionFailed != null)
+            {
+                this.ConnectionFailed(this, new ConnectionFailedEventArgs(e));
+            }
+        }
+
+        //---------------------------------------------------------------------
+        // Publics
+        //---------------------------------------------------------------------
+
+        /// <summary>
+        ///  Create a listener using a dynamically selected, unused local port.
+        /// </summary>
+        public static SshRelayListener CreateLocalListener(ISshRelayEndpoint server)
+        {
+            return CreateLocalListener(server, PortFinder.FindFreeLocalPort());
+        }
+
+        /// <summary>
+        ///  Create a listener using a defined local port.
+        /// </summary>
+        public static SshRelayListener CreateLocalListener(ISshRelayEndpoint server, int port)
+        {
+            if (port < 0 || port > ushort.MaxValue)
+            {
+                throw new ArgumentException("port");
+            }
+
+            return new SshRelayListener(server, port);
+        }
+
+        /// <summary>
+        /// Perpetually listen and relay traffic until cancelled.
+        /// </summary>
+        public Task ListenAsync(CancellationToken token)
+        {
+            return Task.Run(() =>
+            {
+                this.listener.Start(BacklogLength);
+
+                using (token.Register(this.listener.Stop))
+                {
+                    while (this.ClientAcceptLimit == 0 || this.ClientAcceptLimit > this.ClientsAccepted)
+                    {
+                        try
+                        {
+                            var clientStream = new SocketStream(this.listener.AcceptSocket());
+                            var serverStream = new SshRelayStream(this.server);
+
+                            OnClientConnected(clientStream.ToString());
+                            this.ClientsAccepted++;
+
+                            Task.WhenAll(
+                                    clientStream.RelayToAsync(serverStream, token),
```

Comment:
just a question: you're basically piping input->output streams.  What happens on disconnect from either end?  The whole thing terminates or is there a reconnect/retry somewhere i didn't see

Reply:
i guess the safest (easiest thing) to do is make sure each stream is actually terminated (cleanly/closed).  We want the RPD connection closed always

Reply:
also, i wonder if its we could later bind ot a unix-like socket,...i don't know if RDCMan.exe supports binding a file (probably doesnt)

my quick read here indicates we can use file handlers vs actual tcp sockets  https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/

Reply:
The `RelayToAsync` method propagates disconnects. Once both sides have closed the connection,  `Task.WhenAll` returns and the loop starts anew.

So `ListenAsync` will continue accepting and relaying new connections until the `ClientAcceptLimit` is reached or the `CancellationToken` is set. I have been pretty careful to handle connection closes properly, so I hope I properly caught all scenarios.

Regarding Unix sockets -- the limiting factor is the RDP ActiveX. AFAIK, it only supports connecting to an IP/port, not to named pipes or anything else.